### PR TITLE
support block scoped vars captured in closures inside loops

### LINKED ISSUE #3915: Transpile separate scope in each for loop iteration with let and ES5
Hi,

I was recently surprised to see that TypeScript 1.5 (using ntypescript, so recent master) doesn't support transpiling down to ES5 the separate scope functionality of ES6 when using a for loop with a let-initialized iterator.

For example:
```typescript
function buttonsWithLet(count, targetElement) {
  for (let i = 1; i <= count; i+= 1) {  //error 4091:  Loop_contains_block_scoped_variable_0_referenced_by_a_function_in_the_loop_This_is_only_supported_in_ECMAScript_6_or_higher
    var button = makeButton("alert " + i);
    button.onclick = function() {
      alert("This is button " + i + ".");
    };
    targetElement.appendChild(button);
  }
}
```

This is where Error 4091 was implemented.
https://github.com/Microsoft/TypeScript/commit/5f2588f018040f5fcd2a5cb9e31d90fd72b4c597
https://github.com/Microsoft/TypeScript/commit/3b3a94c7d7bea54baec55507d67dc458004da59d

Is the TypeScript team open to transpiling this code in ES5 mode?  If so, would you accept a pull request?  The above code is transpiled by Babel as such which is really clean:

```js
function buttonsWithLet(count, targetElement) {
    var _loop = function (i) {
        button = makeButton("alert " + i);

        button.onclick = function () {
            alert("This is button " + i + ".");
        };
        targetElement.appendChild(button);
    };

    for (var i = 1; i <= count; i += 1) {
        var button;

        _loop(i);
    }
}
```

but a plausibly easier transpile might be:

```js
function buttonsWithVarAndIIFE(count, targetElement) {
    for (var i = 1; i <= count; i += 1) {
        var button = makeButton("alert " + i);
        button.onclick = (function (i) {
            return function () {
                alert("This is button " + i + ".");
            };
        })(i);
        targetElement.appendChild(button);
    }
}
```

### PR DESCRIPTION
**Pull request by vladima**
Fixes #3915

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I noticed that there isn't a flag to enable this. We should consider having a flag to warn when block-scoped variables trigger a capture, or create a TSLint rule for it.

**Comment by vladima** [comment]
Don't think we should hide it under the flag:
- this is a missing bit in our ES6 compliance story that we deliberately decided to punt before.
- this is not a breaking change, code that used to compile before will continue to compile.

Adding TSLint rule for it sounds reasonable

**Comment by vladima** [comment]
pinging @mhegazy: can you have a look at this PR when you have time?

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -379,6 +379,93 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
 
+        const enum Jump {
+            Break       = 1 << 1,
+            Continue    = 1 << 2,
+            Return      = 1 << 3
+        }
+
+        interface ConvertedLoopState {
+            // set of labels that occured inside the converted loop
+            // used to determine if labeled jump can be emitted as is or it should be dispatched to calling code
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Can you use JSDoc for these?

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -379,6 +379,93 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
             return true;
         }
 
+        const enum Jump {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Might as well `JumpKind`.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -379,6 +379,93 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+            // i.e.
+            // for (let x;;) <statement that captures x in closure and uses 'arguments'>
+            // should be converted to
+            // var loop = function(x) { <code where 'arguments' is replaced witg 'arguments_1'> }
+            // var arguments_1 = arguments
+            // for (var x;;) loop(x);
+            // otherwise semantics of the code will be different since 'arguments' inside converted loop body 
+            // will refer to function that holds converted loop.
+            // This value is set on demand.
+            argumentsName?: string;
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: I don't see an example of `arguments` being used both inside and outside of the loop. Can you point it out to me or add one?

**Comment by DanielRosenwasser** [comment]
I don't see an example of `arguments` being used both inside and outside of the loop. Can you point it out to me or add one?

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -2996,7 +3122,293 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                    write(" {");
+                    writeLine();
+                    increaseIndent();
+                }
+
+                const convertedOuterLoopState = convertedLoopState;
+                convertedLoopState = {};
+
+                if (convertedOuterLoopState) {
+                    // convertedOuterLoopState !== undefined means that this converted loop is nested in another converted loop
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: It seems like this line should have a period at the end of it.

File: src/compiler/emitter.ts [hunk 5]
```diff
@@ -2996,7 +3122,293 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                writeLine();
+
+                if (convertedLoopState.argumentsName) {
+                    // if alias for arguments is set
+                    if (convertedOuterLoopState) {
+                        // pass it to outer converted loop
+                        convertedOuterLoopState.argumentsName = convertedLoopState.argumentsName;
+                    }
+                    else {
+                        // this is top level converted loop and
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: "and we need to"

File: src/compiler/emitter.ts [hunk 6]
```diff
@@ -2996,7 +3122,293 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                        write("var ");
+                        let seen: Map<string>;
+                        for (const id of convertedLoopState.hoistedLocalVariables) {
+                           if (!seen) {
+                               seen = {};
+                           }
+                           else {
+                               write(", ");
+                           }
+                           if (!hasProperty(seen, id.text)) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Add a line above

File: src/compiler/emitter.ts [hunk 7]
```diff
@@ -2996,7 +3122,293 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                    if (convertedOuterLoopState) {
+                        // pass them to outer converted loop
+                        convertedOuterLoopState.hoistedLocalVariables = convertedLoopState.hoistedLocalVariables;
+                    }
+                    else {
+                        // deduplicate and hoist collected variable declarations
+                        write("var ");
+                        let seen: Map<string>;
+                        for (const id of convertedLoopState.hoistedLocalVariables) {
+                           if (!seen) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]:
```ts
// Don't initialize seen unless we have at least one element.
// Separate by commas for all but the first element.
```

File: src/compiler/emitter.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Actually `Jump` is fine as long as you just document what these are for (i.e. tracking non-local jumps for local variable capturing in loops)

File: src/compiler/emitter.ts [hunk 8]
```diff
@@ -3195,12 +3628,63 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+
                 emitToken(node.kind === SyntaxKind.BreakStatement ? SyntaxKind.BreakKeyword : SyntaxKind.ContinueKeyword, node.pos);
                 emitOptional(" ", node.label);
                 write(";");
             }
 
             function emitReturnStatement(node: ReturnStatement) {
+                if (convertedLoopState) {
+                    convertedLoopState.nonLocalJumps |= Jump.Return;
+                    write(`return { value: `);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Just use a regular string.

File: src/compiler/emitter.ts [hunk 9]
```diff
@@ -3195,12 +3628,63 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                    if (node.expression) {
+                        emit(node.expression);
+                    }
+                    else {
+                        write("undefined");
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: I can't seem to find any tests for this.

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Also, should it be `void 0`?

**Comment by DanielRosenwasser** [comment]
What about `this` capturing?

**Comment by vladima** [comment]
@DanielRosenwasser ok

File: src/compiler/emitter.ts (hunk 4 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

File: src/compiler/emitter.ts (hunk 5 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

File: src/compiler/emitter.ts (hunk 6 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

File: src/compiler/emitter.ts (hunk 7 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

File: src/compiler/emitter.ts (hunk 9 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

File: src/compiler/emitter.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

File: src/compiler/emitter.ts (hunk 8 above)
Comment (vladima) [inline src/compiler/emitter.ts]: done

**Comment by vladima** [comment]
@DanielRosenwasser any other comments?

File: src/compiler/emitter.ts [hunk 10]
```diff
@@ -3378,10 +3886,28 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 write(";");
             }
 
-            function emitLabelledStatement(node: LabeledStatement) {
+            function emitLabelAndColon(node: LabeledStatement): void {
                 emit(node.label);
                 write(": ");
+            }
+
+            function emitLabelledStatement(node: LabeledStatement) {
```
Comment (RyanCavanaugh) [inline src/compiler/emitter.ts]: Spelling

Comment (vladima) [inline src/compiler/emitter.ts]: oops, thanks

**Comment by mhegazy** [comment]
:+1:

**Comment by nycdotnet** [comment]
This is awesome!!!

**Comment by Nemikolh** [comment]
With version `typescript@1.8.0-dev.20151121` this example:
```ts
class C {
  constructor(private N: number) {}
  foo() {
    for (let i of [0]) {
      let f = () => i;
      this.N = f();
    }
  }
}
```
is not correctly transformed.

**Comment by DanielRosenwasser** [comment]
Thanks for the heads up @Nemikolh. I filed #5746 on your behalf.

**Comment by Nemikolh** [comment]
Glad to help! :)

**Pull request merged**